internal/datanode/buffer.go

```diff
@@ -17,13 +17,220 @@
+		fmt.Sprint(Params.DataNodeCfg.GetNodeID()), metrics.DeleteLabel).Add(float64(rowCount))
+}
+
+func (bm *DelBufferManager) Load(segID UniqueID) (delDataBuf *DelDataBuf, ok bool) {
+	value, ok := bm.delBufMap.Load(segID)
+	if ok {
+		return value.(*DelDataBuf), ok
+	}
+	return nil, ok
+}
+
+func (bm *DelBufferManager) Delete(segID UniqueID) {
+	buf, ok := bm.delBufMap.Load(segID)
+	if ok {
+		delDataBuf := buf.(*DelDataBuf)
+		item := delDataBuf.item
+		bm.delMemorySize -= item.memorySize
+		heap.Remove(bm.delBufHeap, item.index)
+		bm.delBufMap.Delete(segID)
+	}
+}
+
+func (bm *DelBufferManager) LoadAndDelete(segID UniqueID) (delDataBuf *DelDataBuf, ok bool) {
+	buf, ok := bm.Load(segID)
+	if ok {
+		bm.Delete(segID)
+		return buf, ok
+	}
+	return nil, ok
+}
+
+func (bm *DelBufferManager) CompactSegBuf(compactedToSegID UniqueID, compactedFromSegIDs []UniqueID) {
+	var compactToDelBuff *DelDataBuf
+	compactToDelBuff, loaded := bm.Load(compactedToSegID)
+	if !loaded {
+		compactToDelBuff = newDelDataBuf()
+	}
+
+	for _, segID := range compactedFromSegIDs {
+		if delDataBuf, loaded := bm.LoadAndDelete(segID); loaded {
+			compactToDelBuff.mergeDelDataBuf(delDataBuf)
+		}
+	}
+	// only store delBuf if EntriesNum > 0
+	if compactToDelBuff.EntriesNum > 0 {
+		if loaded {
+			bm.delBufHeap.update(compactToDelBuff.item, compactToDelBuff.item.memorySize)
+		} else {
+			heap.Push(bm.delBufHeap, compactToDelBuff.item)
+		}
+		//note that when compacting segment in del buffer manager
+		//there is no need to modify the general memory size as there is no new
+		//added del into the memory
+		bm.delBufMap.Store(compactedToSegID, compactToDelBuff)
+	}
+}
+
+func (bm *DelBufferManager) ShouldFlushSegments() []UniqueID {
+	var shouldFlushSegments []UniqueID
+	if bm.delMemorySize < Params.DataNodeCfg.FlushDeleteBufferBytes {
+		return shouldFlushSegments
+	} else {
+		mmUsage := bm.delMemorySize
+		var popedSegMem []*Item
+		for {
+			segMem := heap.Pop(bm.delBufHeap).(*Item)
+			popedSegMem = append(popedSegMem, segMem)
+			shouldFlushSegments = append(shouldFlushSegments, segMem.segmentID)
+			mmUsage -= segMem.memorySize
+			if mmUsage < Params.DataNodeCfg.FlushDeleteBufferBytes {
+				break
+			}
+		}
+		//here we push all selected segment back into the heap
+		//in order to keep the heap semantically correct
+		for _, segMem := range popedSegMem {
+			heap.Push(bm.delBufHeap, segMem)
+		}
+		return shouldFlushSegments
+	}
+}
+
+// An Item is something we manage in a memorySize priority queue.
+type Item struct {
+	segmentID  UniqueID // The segmentID
+	memorySize int64    // The size of memory consumed by del buf
+	index      int      // The index of the item in the heap.
+	// The index is needed by update and is maintained by the heap.Interface methods.
+}
+
+// A PriorityQueue implements heap.Interface and holds Items.
+// We use PriorityQueue to manage memory consumed by del buf
+type PriorityQueue []*Item
```

Comment:
there is already a `PriorityQueue` in `indexcoord`, should move these 2 types in util and reuse